File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] arr = new int[nums.length];
        Arrays.fill(arr, -1);

        Stack<Integer> stack = new Stack<>();

        for(int i=0; i<2*nums.length; i++){
            int index = i % nums.length;
            while(!stack.isEmpty() && nums[stack.peek()] < nums[index]){
                arr[stack.pop()] = nums[index];
            }

            if(i < nums.length) stack.push(i);
        }

        return arr;
    }
}

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> arr(nums.size(), -1);
        stack<int> stack;

        for(int i=0; i<2*nums.size(); i++){
            int index = i%nums.size();
            while(!stack.empty() && nums[stack.top()] < nums[index]){
                int topIndex = stack.top();
                arr[topIndex] = nums[index];
                stack.pop();
            }

            if(i < nums.size()) stack.push(i);
        }

        return arr;
    }
};